s2/gen.go

```diff
@@ -0,0 +1,1590 @@
+//+build generate
+
+//go:generate go run gen.go -out encodeblock_amd64.s -stubs encodeblock_amd64.go
+
+package main
+
+import (
+	"fmt"
+	"log"
+
+	. "github.com/mmcloughlin/avo/build"
+	"github.com/mmcloughlin/avo/buildtags"
+	"github.com/mmcloughlin/avo/operand"
+	. "github.com/mmcloughlin/avo/operand"
+	"github.com/mmcloughlin/avo/reg"
+)
+
+func main() {
+	Constraint(buildtags.Not("appengine").ToConstraint())
+	Constraint(buildtags.Not("noasm").ToConstraint())
+	Constraint(buildtags.Term("gc").ToConstraint())
+
+	genEncodeBlockAsm("encodeBlockAsm", 16, 6, false)
+	genEncodeBlockAsm("encodeBlockAsm14B", 14, 5, false)
+	genEncodeBlockAsm("encodeBlockAsm12B", 12, 4, false)
+	genEncodeBlockAsm("encodeBlockAsmAvx", 16, 6, true)
+	genEncodeBlockAsm("encodeBlockAsm14BAvx", 14, 5, true)
+	genEncodeBlockAsm("encodeBlockAsm12BAvx", 12, 4, true)
+	genEmitLiteral()
+	genEmitRepeat()
+	genEmitCopy()
+	genMatchLen()
+	Generate()
+}
+
+func debugval(v operand.Op) {
+	value := reg.R15
+	MOVQ(v, value)
+	INT(Imm(3))
+}
+
+func genEncodeBlockAsm(name string, tableBits, skipLog int, avx bool) {
+	TEXT(name, 0, "func(dst, src []byte) int")
+	Doc(name+" encodes a non-empty src to a guaranteed-large-enough dst.",
+		"It assumes that the varint-encoded length of the decompressed bytes has already been written.", "")
+	Pragma("noescape")
+
+	// "var table [maxTableSize]uint32" takes up 4 * (1 << tableBits) bytes of stack space.
+	// Extra bytes are added to keep less used values.
+	var (
+		tableSize = 1 << tableBits
+		// Keep base stack multiple of 16.
+		baseStack = 0
+		// try to keep extraStack + baseStack multiple of 16
+		// for best chance of table alignment.
+		extraStack = 32
+		allocStack = baseStack + extraStack + tableSize
+	)
+
+	// Memzero needs at least 128 bytes.
+	if tableSize < 128 {
+		panic("tableSize must be at least 128 bytes")
+	}
+
+	lenSrcBasic, err := Param("src").Len().Resolve()
+	if err != nil {
+		panic(err)
+	}
+	lenSrcQ := lenSrcBasic.Addr
```

Comment:
I'm wondering why you are doing it this way instead of the following?

```go
lenSrc := Load(Param("src").Len(), GP64())
```

Reply:
That would use a register. With this I only have a reference in avo, so I can use `lenSrcQ` to look up the value on the stack.